primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use crate::v3::{BlakeTwo256, HashT as _};
+use parity_scale_codec::{Decode, Encode};
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// A single executor parameter
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutorParam {
+	/// ## General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
+
+	/// ## Parameters setting the executuion environment semantics:
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit (max. number of Wasm values on stack)
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+
+	/// Placeholder for any data not fitting into the semantics. It may be used to avoid
+	/// `SessionInfo` migration if fast adoption of a new parameter is needed. This is an
+	/// emergency measure, only use it if you ought to.
+	#[codec(index = 255)]
+	RawData(Vec<u8>),
```

Comment:
@eskimor, I need some consultation here.

I remember we agreed that we're introducing such a parameter to mitigate the influence of "unknown unknowns" and to be able to introduce a new parameter without a proper migration if needed to be done asap and do a proper migration afterward. But reviewing this code now, I stopped understanding how it could help.

For example, suppose we find out that we need control over some previously unknown `bool` parameter asap. In that case, we encode that `bool` to `Vec<u8>` and add some code to the executor interface to decode it and tune the execution environment accordingly. But we _still_ need to wait for the majority of nodes to upgrade to be able to use this parameter. So there's no win over a regular migration. Did I miss/forget something?

Reply:
For a proper migration, we need to upgrade the runtime as well. So it will be more lengthy. But I really don't have a strong opinion here.

Reply:
> For a proper migration, we need to upgrade the runtime as well.

We need it to set the `RawData` value as well. It is still set in the `session_info` pallet code. So looks like a non-working solution :man_shrugging:

Reply:
Setting the value can be done via governance without a runtime upgrade. But I agree, it is probably not worth it.

Reply:
Then the node implementation would still need to be able to interpret this.